### PR TITLE
update app.js (HNY)

### PR DESCRIPTION
**Pull request by AlinaLadybug**
### Object-Oriented JavaScript

[Demo](https://alinaladybug.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/AlinaLadybug/frontend-nanodegree-arcade-game/tree/master/js)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/AlinaLadybug/object-oriented/app.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
+const CELL = { width: 101, height: 83 };
+const CANVAS = { width: 505, height: 606 };
+const ENEMY = { speed: 100, startPosition: -50, width: 80, height: 60 };
+const PLAYER_LOCATION = { x_axis: 200, y_axis: 400 };
```
Comment (OleksiyRudenko) [inline submissions/AlinaLadybug/object-oriented/app.js]: Please, format the literals as per conventions, one line per property+value. This take some extra lines of code but also adds to readability.

File: submissions/AlinaLadybug/object-oriented/app.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+const Creature = function (x, y) {
+    this.x = x;
+    this.y = y;
```
Comment (OleksiyRudenko) [inline submissions/AlinaLadybug/object-oriented/app.js]: You have introduced a base class to hold shared properties `x` and `y`.
Please, note that `sprite` is also a shared property and `render()` is a shared method. Please, move those to `Creature`.

File: submissions/AlinaLadybug/object-oriented/app.js (hunk 1 above)
Comment (OleksiyRudenko) [reply submissions/AlinaLadybug/object-oriented/app.js]: Per convention underscore is used to delimit parts of all caps identifiers like `PLAYER_LOCATION`.
When any identifiers aren't all caps, use camelCase. e.g. `xAxis`, like `startPosition`.

File: submissions/AlinaLadybug/object-oriented/app.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    if (this.x > CANVAS.width) {
+        this.x = ENEMY.startPosition;
+        this.speed = Math.floor(Math.random() * 200) + ENEMY.speed;
+    }
+
+    if (player.x > this.x - ENEMY.width && player.x < this.x + ENEMY.width &&
+        player.y > this.y - ENEMY.height && player.y < this.y + ENEMY.height) {
+        player.x = PLAYER_LOCATION.x_axis;
+        player.y = PLAYER_LOCATION.y_axis;
+    }
```
Comment (OleksiyRudenko) [inline submissions/AlinaLadybug/object-oriented/app.js]: This block seem to handle collisions. As this is an important feature it is worth being moved to its own dedicated method to follow the Separation of Concerns principle.

Also `player` is a global variable. Classes should be agnostic of the outer world. The only globals allowed are constants and library functions.

Please, find a way for enemies to know the player explicitly.

File: submissions/AlinaLadybug/object-oriented/app.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+let allEnemies = [];
+let player = new Player(PLAYER_LOCATION.x_axis, PLAYER_LOCATION.y_axis);
+let enemiesLocation = [52, 132, 212];
+enemiesLocation.forEach(elem => {
+    let enemy = new Enemy(0, elem, ENEMY.speed);
+    allEnemies.push(enemy);
+})
```
Comment (OleksiyRudenko) [inline submissions/AlinaLadybug/object-oriented/app.js]: Consider the following example:
```
let allEnemies = [52, 132, 212]
   .map(locationY => new Enemy(0, locationY, ENEMY.speed);
```
Which saves you 66% of lines used before and also geds rid of an unnecessary variable.

Of course `new Enemy(...)` will be supplemented with an extra argument, which is a reference to the player.

Beside that `[52, 132, 212]` are also magic numbers worth being constants. BTW, aren't those calculable from `CELL.height`?
`[1,2,3].map(row => row * CELL.height - CELL.yOffset)...`
where `CELL.yOffset` equals `-31` or so.

**Comment by AlinaLadybug** [comment]
@OleksiyRudenko  Thanks for your comments!)

**Pull request merged**